Generic IO exported interface, let declaration fix

tvquizphd opened the pull request:
- ~I've removed two unused declarations detected by `npm run lint`.~
- ~I've used [function overloading][OVERLOAD] on the `dummy_socket` to allow passing tests.~
- ~I've updated the version to `0.4.0-beta.1.0`, with hope to one day release somewhere as 0.4.0.~

EDIT: this PR now merely:
- removes type assertions in tests with a more generic `IO` interface.
- replaces two `let` declarations, and adds `npm run lint` support.
- bumps "libsodium-wrappers-sumo" to `0.7.8`.

[OVERLOAD]: https://www.typescriptlang.org/docs/handbook/declaration-files/by-example.html#overloaded-functions

tvquizphd commented:
Oh— were the tests passing before https://github.com/nthparty/opaque/pull/7/commits/e32d2cb6cc24c773e13f55f2e8de26f7e4046548 ?

AverageHelper commented:
Great catch! You're right, I neglected to re-run the tests after e32d2cb. I think the test failure can be fixed without so many changes to the JavaScript-land code, and I have a smol commit in the works for that. (EDIT: See f7f416d)

My goal with this fork was to make as few changes as possible to add basic TypeScript structures to the existing logic. I figured that few changes to the non-type code increases the chance that this work would be merged. I want also to add strict ESLint and Prettier configs, enforce `const` where `let` is unneeded, etc., but in the interest of minimal changes, I want those to be future enhancements for after the foundation is in place.

Thanks for the help! I'll leave your PR open for now, until @wyatt-howe merges the fork upstream. If they do not act on my PR for some time, then I'll prepare to deploy my fork as its own package, and look into accepting your PR into that.

AverageHelper commented:
Come to think of it, I did make some minor fixes to JS-land code in this fork (`==` to `===` for example), so I think I'd be okay with other such minor fixes (`let` to `const` for example). Anything larger, like adding ESLint, should probably happen in a separate PR to the TS port I think.

As for the unused parameters, I don't know the OPAQUE spec at all (yet), so I don't know that those parameters weren't intended to be used. If those parameters *should* have been used according to the spec, I'd like ESLint (later) to point them out so future maintainers can spot them to check up on why those params were included, and whether the spec calls for their use somehow.

AverageHelper commented on package.json:
```diff
@@ -1,13 +1,13 @@
 {
   "name": "@nthparty/opaque",
-  "version": "0.3.0",
+  "version": "0.4.0-beta.1.0",
```
Might be best to let the upstream maintainers decide what version this winds up being, likely in a separate PR as per [their usual pattern](https://github.com/nthparty/opaque/commits/main). There aren't any runtime changes here as far as I know, so I expect this would only be a semver-minor bump. Their deployment schedule might call for a beta tag anyway tho. I'd like not to make that decision for them unless I decide to deploy this fork to NPM separately.

AverageHelper commented on package.json:
```diff
@@ -1,13 +1,13 @@
     "build": "tsc -p tsconfig.prod.json && cp -r src/types dist/types",
     "test": "jest --coverage",
-    "coveralls": "npm test && ./node_modules/coveralls/bin/coveralls.js < coverage/lcov.info"
+    "lint": "eslint src"
```
I might have missed moving the `coveralls` script elsewhere, but it seems to me that it was deleted. If so, then it may be best to uninstall the `coveralls` package as well.

That said, .travis.yml references this script, so maybe leave it in?

AverageHelper commented on package.json:
```diff
@@ -50,13 +50,15 @@
-    "libsodium-wrappers-sumo": "0.7.6",
-    "oprf": "2.0.0"
+    "libsodium-wrappers-sumo": "0.7.8",
+    "oprf": "^2.0.0"
```
This package should be reinstalled with the `-E` option. My aim with pinning exact versions generally is to avoid supply-chain attacks (such as the sort [proliferated by `node-ipc` not long ago](https://snyk.io/blog/peacenotwar-malicious-npm-node-ipc-package-vulnerability/)). The same should probably be done with `eslint` and `"@typescript-eslint/eslint-plugin` for consistency, if those packages are to remain.

AverageHelper commented on .eslintrc.json:
```diff
@@ -0,0 +1,17 @@
+  "extends": [
+    "eslint:recommended",
+    "plugin:@typescript-eslint/recommended"
+  ],
+  "rules": {
+    "@typescript-eslint/no-explicit-any": "off"
+  }
```
Is there any case for explicit `any` that shouldn't instead use `unknown` or a more specific type?

```suggestion
  ]
```

tvquizphd commented:
I've rebased this PR after https://github.com/AverageHelper/opaque/commit/f7f416d6450a8e2a83ef8e4264ea2eeed2957a68.

Now all this PR does is:
- Add `npm run lint` with eslint dependency
- Replace two `let` declarations with `const`

I've reverted my typescript changes and removal of unused variables.

tvquizphd commented:
> Thanks for the help! I'll leave your PR open for now, until @wyatt-howe merges the fork upstream. If they do not act on my PR for some time, then I'll prepare to deploy my fork as its own package, and look into accepting your PR into that.

This works well for me. I'm planning on modifying my own fork further, but I hope @wyatt-howe merges [your PR](https://github.com/nthparty/opaque/pull/7)!

tvquizphd commented:
I've removed the `eslint` dependency. More importantly, This PR removes type assertions in the tests with a more generic `IO` interface. This is the solution I've settled on with [my own fork][my-fork]. Though, on my fork I also have runtime checks to ensure only expected tags and values are sent or accepted through `IO`.

[my-fork]: https://github.com/tvquizphd/opaque#readme

AverageHelper commented:
whoops, didn't think deleting the branch would close this! I think the history there is pretty much what was in the upstream since https://github.com/nthparty/opaque/pull/7 (i might have some other changes on one of my old computers somewhere but it'll be a while before i get around to pushing those, if they exist :sweat_smile:)

if you wanna rebase this work against the main branch, this can be reopened!
